sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryPlanner.java

```diff
@@ -77,7 +87,31 @@ public static LogicalPlan plan(InsertFromSubQueryAnalyzedStatement statement,
         }
         SOURCE_LOOKUP_CONVERTER.process(subRelation, null);
         LogicalPlan plannedSubQuery = logicalPlanner.plan(subRelation, plannerContext, subqueryPlanner, FetchMode.NEVER_CLEAR);
-        return new Insert(plannedSubQuery, indexWriterProjection);
+
+        EvalProjection castOutputs = createCastProjection(statement.columns(), plannedSubQuery.outputs());
+        List<Projection> projections = castOutputs == null
+            ? Collections.singletonList(indexWriterProjection)
+            : ImmutableList.of(castOutputs, indexWriterProjection);
+        return new Insert(plannedSubQuery, projections);
+    }
+
+    @Nullable
+    private static EvalProjection createCastProjection(List<Reference> targetCols, List<Symbol> sources) {
```

Comment:
I don't like the `sources` name here, maybe sourceCols?
`sources` refers for me more to relations...